[MRG] DOC: update somato example

rythorpe opened the pull request:
Goals:

- [x] improve narrative
- [ ] ~~update MNE vertex selection~~
- [x] update drive and local network connectivity parameters
- [x] consolidate plots so that the median nerve response inverse solution (i.e., MNE vertex) and hnn-core simulation are plotted together
- [x] provide option for multiple trials that, when averaged together, obviously resemble the empirical inverse solution
- [x] Extra fix: allow sync_within_trials to be set from within `Network` constructor if `add_drives_from_params is True`

closes #247

jasmainak commented:
let me know when I should look!

rythorpe commented:
@jasmainak any idea why I seem to produce an MNE vertex time course (shown below) that looks different from the [one on our documentation page](https://jonescompneurolab.github.io/hnn-core/auto_examples/plot_simulate_somato.html#sphx-glr-auto-examples-plot-simulate-somato-py)?

![Screenshot from 2021-01-25 23-11-01](https://user-images.githubusercontent.com/20212206/105801200-a1c0c480-5f66-11eb-98c6-bd80bbd9f94f.png)

rythorpe commented on hnn_core/param/N20.json:
```diff
@@ -1,5 +1,5 @@
 {
-	"tstop": 120,
+	"tstop": 170,
```
I know it's not ideal to make the simulations we run in our examples longer, however, this is to maintain consistency with `default.json` and the work that will (hopefully) get published soon where we explicitly model the median nerve evoked response.

rythorpe commented on examples/plot_simulate_somato.py:
```diff
@@ -3,20 +3,21 @@
 05. Source reconstruction and HNN
```
This might be a little wordy, but what do you think of `05. From sensor to simulation: MEG source reconstruction with HNN` so that the title is more descriptive?

jasmainak replied:
+1 for more descriptive title. However, "MEG source reconstruction with HNN" is a bit inaccurate ... because HNN does not do source reconstruction. Maybe:

05. From MEG sensor-space data to HNN simulation

? Any other title that's not inaccurate will do :)

rythorpe replied:
I'm not sure I understand what is inaccurate about it. "Source reconstruction" is what we had originally.

cjayb replied:
I think @jasmainak means it implies that `hnn_core` is doing the reconstructing. I'm not too keen on the 'From' though, since these are really two completely separate code sequences tacked after one another. How about?

```suggestion
05. Comparing MEG source reconstruction to HNN model prediction
```

rythorpe replied:
Sorry, I just noticed the most recent comment now after reloading page. I actually don't think of the MNE inverse solution and HNN simulation as "completely separate code sequences" at all. In fact, very much the opposite. What we're trying to demonstrate is a process that most users won't find intuitive. Fitting sensor-level data into the HNN framework is a process that requires interdependent considerations between the inverse solution technique and HNN, and the only reason it appears simple in this example is because we've left out the narrative about how I got the specific results and values that we present in this example. I'm not committed to any title in particular, although @jasmainak's suggestion is growing on me. Big picture: we should find a way to communicate that this is a process rather than an end result to the user.

jasmainak commented:
> any idea why I seem to produce an MNE vertex time course (shown below) that looks different from the one on our documentation page?

different MNE version :-) I don't think we can really guarantee exactly the same result. But it might be worthwhile to check if the source localization looks correct by doing `stc.plot` and then picking a vertex manually using the pyvista backend. I wanted to do this myself but for some reason the renderer is broken on my mac ...

rythorpe commented:
> different MNE version :-) I don't think we can really guarantee exactly the same result. But it might be worthwhile to check if the source localization looks correct by doing `stc.plot` and then picking a vertex manually using the pyvista backend. I wanted to do this myself but for some reason the renderer is broken on my mac ...

After we resolve this, it might be a good idea to plot the dipole in source space for this example to show how we interpret orientation in the HNN framework. Is the an efficient way to do this that doesn't require a fancy graphics backend?

jasmainak commented:
> Is there an efficient way to do this that doesn't require a fancy graphics backend?

MNE does have a 'matplotlib' backend for plotting stc. It might be pretty slow though ... and neglected. OpenGL makes it fast. Speed comes at a cost ... always :-P

jasmainak commented:
see here the plot: https://mne.tools/stable/auto_tutorials/source-modeling/plot_visualize_stc.html#volume-source-estimates just above "Volume Source Estimates"

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +103,90 @@
+# match to the empirical waveform, set ``n_trials`` to be >=25.
+n_trials = 2
+# n_trials = 25
+with MPIBackend(n_procs=6, mpi_cmd='mpiexec'):
```
I think better to set this to `n_procs=1`? What would happen if some computer doesn't have 6 cores?

rythorpe replied:
Good point. How about `n_procs=2` so that the backend at least does something?

cjayb replied:
+1 for `n_procs=2`, dual-core is standard since way back when.

FYI I ran the example on my dual-core i7 with `n_procs=6` and it worked without flaf! Of course each `nrniv` process was down to 33% as the CPU was getting choked on thread hopping :)

cjayb replied:
But do we really need the explicit `mpi_cmd=`? I think it'll confuse newcomers more than it helps power-users.

rythorpe replied:
Sounds good. Updated in https://github.com/jonescompneurolab/hnn-core/pull/253/commits/3f665c45e7e76eb9d9282fc5ba025196f3368453.

rythorpe commented:
> see here the plot: https://mne.tools/stable/auto_tutorials/source-modeling/plot_visualize_stc.html#volume-source-estimates just above "Volume Source Estimates"

Can the matplotlib backend render a vector inverse solution?

jasmainak commented on hnn_core/network.py:
```diff
@@ -248,6 +248,8 @@ def __init__(self, params, add_drives_from_params=False,
                         sigma=specs['dynamics']['sigma'],
                         numspikes=specs['dynamics']['numspikes'],
+                        sync_within_trial=specs['dynamics']
+                                               ['sync_within_trial'],
```
Is this a bugfix?

rythorpe replied:
Correct, this is for #247.

cjayb replied:
LGTM, thanks!

jasmainak commented:
> Can the matplotlib backend render a vector inverse solution?

honestly don't know, you just have to try ...

cjayb commented on hnn_core/param/N20.json:
```diff
@@ -8,7 +8,7 @@
 	"f_max_spec": 40.0,
 	"dipole_scalefctr": 30,
-	"dipole_smooth_win": 5,
+	"dipole_smooth_win": 20,
```
Opened a new issue #262 to get rid of these hidden, plot-affecting paramters

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +103,90 @@
+fig, axes = plt.subplots(3, 1, sharex=True, figsize=(6, 6))
+net.cell_response.plot_spikes_hist(ax=axes[0], show=False)
+axes[0].legend(['evdist_1', 'evdist_2', 'evprox_1', 'evprox_2'])
+axes[0].set_ylabel('Counts')
```
Would you please add this to `viz.plot_spikes_hist`? It's clearly needed everywhere!

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +103,90 @@
+# nerve evoked response waveforms, and output spike histogram.
+fig, axes = plt.subplots(3, 1, sharex=True, figsize=(6, 6))
+net.cell_response.plot_spikes_hist(ax=axes[0], show=False)
+axes[0].legend(['evdist_1', 'evdist_2', 'evprox_1', 'evprox_2'])
```
Hmm, this is an odd hack, but I see it has to do with some `matplotlib`-intricacies when doing subplots...

ntolley replied:
I'll need to test but this may be easily fixed from the function side.

rythorpe replied:
See https://github.com/jonescompneurolab/hnn-core/pull/253/commits/7dcb64532c5aa3e75fc3fddf7f8f54da2df68554.

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +103,90 @@
+axes[0].legend(['evdist_1', 'evdist_2', 'evprox_1', 'evprox_2'])
+axes[0].set_ylabel('Counts')
+axes[1].axhline(0, c='k', ls=':')
+axes[1].plot(1e3 * stc.times, stc.data[pick_vertex, :].T * 1e9, 'r--')
```
Add a legend to the plot here too? 'MNE' and 'HNN', I guess?

rythorpe replied:
Done. I put 'MNE vertex' and 'HNN simulation' to be explicit.

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -3,20 +3,21 @@
-somatosensory dataset.
+This example demonstrates how to calculate the inverse solution of the median
+nerve evoked response in the MNE somatosensory dataset, and then simulate a
+matching inverse solution with HNN.
```
```suggestion
biophysical model network that reproduces the observed dynamics.
```

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,37 @@
-# to note that the dipole currents simulate with HNN are assumed to be normal
+# to note that the dipole currents simulated with HNN are assumed to be normal
 # to the cortical surface. Hence, using the option ``pick_ori='normal'``
 # seems to make most sense.
```
```suggestion
is appropriate.
```

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,37 @@
 
+###############################################################################
+# We isolate and plot the single most active vertex in the distributed minimum
+# norm estimate by calculating the L2 norm of the time course emerging from
```
```suggestion
# norm estimate by calculating the L2 norm of the time course at
```

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,37 @@
 
 import hnn_core
-from hnn_core import simulate_dipole, read_params, Network
+from hnn_core import simulate_dipole, read_params, Network, MPIBackend, average_dipoles
```
PEP8

cjayb commented:
This is fantastic stuff @rythorpe !! You should rebase to get @ntolley 's mod to the spike raster. I tried to simplify the plotting sections, but I see why you've done what you have. This is an advanced demo, so more explicit python is ok imho. So cool!

jasmainak reviewed (approved):
PR looks fantastic. +1 to merge once you have addressed @cjayb 's comments about matplotlib. The green button is yours @cjayb

rythorpe commented:
I think I've addressed all of the comments above, except for our discussion about the title (@cjayb lmk what you think, I'm happy to move forward with any of the titles suggested thus far) and our internal sanity check of plotting the MNE dipole on the structural MRI. @jasmainak I played around with plotting the MNE and encountered a few things that are a bit confusing. Would you be willing to have a short meeting before we merge this PR?

jasmainak commented:
> @jasmainak I played around with plotting the MNE and encountered a few things that are a bit confusing. Would you be willing to have a short meeting before we merge this PR?

I have time tomorrow. Any time after 10 am! Just send me an invite on google calendar

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +104,92 @@
+# nerve evoked response waveforms, and output spike histogram.
+fig, axes = plt.subplots(3, 1, sharex=True, figsize=(6, 6))
+net.cell_response.plot_spikes_hist(ax=axes[0],
+                                   spike_types=['evprox1', 'evdist1',
+                                                'evdist2', 'evprox2'],
```
```suggestion
                                   spike_types=['evprox', 'evdist']
```

I think this will plot the proximal in one color and distal in another instead of 4 different colors

cjayb replied:
It does (+1)

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -1,22 +1,23 @@
 =================================
-05. Source reconstruction and HNN
+05. From MEG sensor-space data to HNN simulation
 =================================
```
Heading markers (`=`) cover entire title

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,38 @@
 plt.show()
 
 ###############################################################################
 # Now, let us try to simulate the same with hnn-core. We read in the network
```
```suggestion
# Now, let us try to simulate the same with ``hnn-core``. We read in the network
```

cjayb reviewed (commented):
Just two nitpicks. Add [MRG] to the title and this'll be in in a jiffy!

cjayb commented:
> I'm happy to move forward with any of the titles suggested thus far) 

I like the present title.

> and our internal sanity check of plotting the MNE dipole on the structural MRI. @jasmainak I played around with plotting the MNE and encountered a few things that are a bit confusing.

Good idea. I haven't brought this up before because I think this example has plenty of merit as it is. However, choosing the maximum vertex of an L2 solution can only act as a sort of face validity test. For physiological insight, we need to think about what that extracted waveform actually represents. In this dataset, the contralateral S1 response seems to be followed by a later, ipsilaterally dominant S2. The L2 vertex represent some mixture of those, so modelling it with HNN hardly makes sense. Which is exactly why it's so exciting to develop `hnn_core` towards multiple simultaneously active patches!

For future reference, a more principled approach might be to use fixed dipoles. Here's some code to illustrate my point:

```python
evoked.plot_topomap(times=[0.0, 0.04, 0.06, 0.10, 0.140])

from mne.dipole import fit_dipole
sphere = mne.make_sphere_model(r0='auto', head_radius='auto', info=raw.info)
evo_170 = evoked.copy().crop(tmin=0.0, tmax=0.170)
dip, _ = fit_dipole(evo_170, cov, sphere)
dip.plot_amplitudes()

best_idx = np.argmax(dip.gof)
best_time = dip.times[best_idx]

# fixed position & orientation
dip_fixed = mne.fit_dipole(evo_170, cov, sphere,
                           pos=dip.pos[best_idx], ori=dip.ori[best_idx])[0]

dip_fixed.plot()
dip.plot_locations(fwd['mri_head_t'], '01', idx=best_idx, show_all=False,
                   subjects_dir=subjects_dir, mode='orthoview')
```

Dipoles, just like aggregate L2-norm solutions come in physiologically relevant units (10's of nAm).

Going forward, we might want to find an example in which there's a clearer contralateral 20-35-60 ms response complex in S1. A single dipole would model that pretty well, and _that_ would be fun to compare to an HNN model :)

jasmainak commented:
I agree dipole fit makes more sense, will try your code when zooming with Ryan to see what aspects can be incorporated

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -181,8 +181,7 @@
 net.cell_response.plot_spikes_hist(ax=axes[0],
-                                   spike_types=['evprox1', 'evdist1',
-                                                'evdist2', 'evprox2'],
+                                   spike_types=['evprox', 'evdist']
```
missing comma at EOL

rythorpe commented:
> Going forward, we might want to find an example in which there's a clearer contralateral 20-35-60 ms response complex in S1. A single dipole would model that pretty well, and that would be fun to compare to an HNN model :)

Let's definitely continue this conversation in later PRs as we improve this example. A single dipole model would be great to demo as well, however, selecting a vertex from a distributed inverse solution has its merits particularly when the true number of sources is unknown (as is the case for even the contralateral response).

On a slightly different note, we may have found a bug or two in the MNE source code so let's hold off on merging just yet until they can be resolved.

rythorpe commented:
@jasmainak @cjayb what do you think of https://github.com/jonescompneurolab/hnn-core/pull/253/commits/be31a9f91407554b9ffb204ba079f9d0c798e998? If it makes things too complicated, we can always remove this commit but it implements the following changes:

1. use dSPM instead of MNE to find the vertex of maximal activity, then extract the time course from the MNE
2. added code + comments to give uses the option to plot the brain activity on a structural MRI

jasmainak commented:
@rythorpe I think it's a nice workflow you've found but my inclination is to hardcode the vertex and use the matplotlib backend to show which vertex has been picked. This will keep the story simple. @cjayb what is your opinion?

I wonder if averaging in a small label with activation would make more sense than picking a single vertex? I guess this goes back to the dipole fit vs MNE discussion ...

cjayb commented:
> @rythorpe I think it's a nice workflow you've found but my inclination is to hardcode the vertex and use the matplotlib backend to show which vertex has been picked. This will keep the story simple. @cjayb what is your opinion?

I agree with @jasmainak : I see your point and your logic to use dSPM is sound @rythorpe , but I feel it takes too much focus. All that jazz just to pick a single vertex... Let's assume many HNN newcomers won't be too versed in intricacies of the inverse problem. For them, you having picked a "representative location" in the hand area should suffice, right?  I'd like to see the hardcoded way---which matplotlib plot are we talking about here?

> I wonder if averaging in a small label with activation would make more sense than picking a single vertex? I guess this goes back to the dipole fit vs MNE discussion ... it's doing some kind of smoothing if you average

Indeed: since you deem the position 'sensible' based on it being in the hand area, how about [extracting the PCA-flipped time course there](https://mne.tools/stable/auto_examples/inverse/plot_label_source_activations.html#sphx-glr-auto-examples-inverse-plot-label-source-activations-py)? I might vote myself down on this, though, as it does add complexity and gives little-to-nothing in terms of how good this example looks!

jasmainak commented:
> which matplotlib plot are we talking about here?

I mean an stc plot but using matplotlib instead of pyvista or mayavi backend. It won't be very much interactive but folks who have full MNE install can change the backend easily.

jasmainak commented:
yeah [pca_flip or mean_flip](https://mne.tools/stable/generated/mne.VectorSourceEstimate.html#mne.VectorSourceEstimate.extract_label_time_course) so you take into account the change in sign due to orientation

jasmainak commented:
yes I noticed that too ;-) @agramfort you need power users like us :)

rythorpe commented:
Okay @jasmainak and @cjayb, what do you think of https://github.com/jonescompneurolab/hnn-core/pull/253/commits/c736f7e439474c17d2670843569636b6b442ee2e? I need to fix the narrative the pca_flip works pretty well.

rythorpe commented:
> returns a `Figure` instance, not `mne.viz.Brain`, so the subsequent `brain.add_foci` bombs :( I'll see if I'm able to find a way to include the `add_foci` as an extra set of `kwargs` (they already have `add_data`).

Same with trying to add an annotation label :/

agramfort commented:
arff these MNE developers that are just good to create bugs 😄  

we need power users but also power contributors ;)

cjayb commented:
> Same with trying to add an annotation label :/

Yeah, it's the same underlying code... I'm not sure out efforts are best spent trying to figure it out. But will the CIs manage the `PyVista` backend, I wonder?

rythorpe commented:
> EDIT: There's my answer... :( I do think we should stick to `mpl` in `hnn_core` for now. The fact that we can `pip install mne` is worth it, even if we don't get the pretty 3D plots in...

Yeah, I'm working on updating our call to the plotting function as well as the commenting as we speak.

cjayb commented:
My mne MPL fix is 8811, in case you want to just plonk it in.

rythorpe commented on examples/plot_simulate_somato.py:
```diff
@@ -4,8 +4,8 @@
 This example demonstrates how to calculate the inverse solution of the median
-nerve evoked response in the MNE somatosensory dataset, and then simulate a
-biophysical model network that reproduces the observed dynamics.
+nerve evoked response in S1 from the MNE somatosensory dataset, and then
```
Added "S1 from" to specify that our method now extracts the evoked response specifically from S1.

rythorpe commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,66 @@
+                    verbose=True)
 
-pick_vertex = np.argmax(np.linalg.norm(stc.data, axis=1))
+###############################################################################
+# To extract the primary response in primary somatosensory cortex (S1), we
+# create a label for the postcentral gyrus (S1) in source-space
+hemi = 'rh'
+label_tag = 'G_postcentral'
+label_s1 = mne.read_labels_from_annot(subject, parc='aparc.a2009s', hemi=hemi,
+                                      regexp=label_tag,
+                                      subjects_dir=subjects_dir)[0]
+
+###############################################################################
+# Visualizing the distributed S1 activation in reference to the geometric
+# structure of the cortex (i.e., plotted on a structural MRI) can help us
+# figure out how to orient the dipole. Note that in the HNN framework,
+# positive and negative deflections of a current dipole source correspond to
+# upwards (from deep to superficial) and downwards (from superficial to deep)
+# current flow, respectively.
+stc_label = stc.in_label(label_s1)
+brain = stc_label.plot(subjects_dir=subjects_dir, hemi='rh', surface='white',
+                       smoothing_steps='nearest', view_layout='horizontal',
+                       initial_time=0.04, backend='matplotlib')
+
+# Uncomment the line below to render a border around the selected label for S1.
+# Note that this requires that you set `backend='pyvista'` above.
+# brain.add_label(label_s1, borders=True)
+
+###############################################################################
+# Now we extract the pca-flipped time course from S1. Note that the most
+# prominent component of the median nerve response occurs in the posterior wall
+# of the central sulcus at ~0.040 sec. Since the dipolar activity here is
+# negative, we orient the extracted waveform so that the deflection at ~0.040
+# sec is pointed downwards.
+flip_data = stc.extract_label_time_course(label_s1, inv['src'],
+                                          mode='pca_flip')
+dipole_tc = -flip_data[0] * 1e9
```
Does this narrative make sense?

rythorpe commented:
@jasmainak @cjayb I've updated the documentation to reflect our changes to the dipole selection process and also updated the HNN params so that the simulated dipole fits better. Give it a look over, but after that I'm good to go.

rythorpe commented:
Looks like CircleCI ran into trouble with importing nibabel when we create the brain plot with the matplotlib backend. Any ideas for what we should do about that?

jasmainak commented:
oh dear ... I didn't realize that `nibabel` was also a dependency for the matplotlib backend. Looks like these things are not really documented either. E.g., [here it's not even listed which backends are allowed](https://mne.tools/dev/generated/mne.viz.use_3d_backend.html). So certainly some fixes could be done there. Fixes in the form of PRs are welcome as always :)

now question is whether we should still stick to this backend. I'm beginning to incline towards `pyvista`. It will probably install `vtk` as well but hopefully that's about it and it's just a pip install.

rythorpe commented:
+1 for `pyvista`. Is this something we'd add to CircleCI or should we just comment out the brain plotting code and let users uncomment it if they wish?

jasmainak commented:
let's add it to CircleCI install script (but not to requirements.txt or anywhere else because we run into the danger of introducing a dependency silently). And we can be explicit to the users that they need the pyvista installed for this example to run

jasmainak commented:
actually looking at the CircleCI log, maybe [it's not so straightforward](https://github.com/mne-tools/mne-python/blob/main/.circleci/config.yml#L55-L66). Okay back to matplotlib. Or maybe we can for skip the plot altogether and leave it as a comment like you did before. Whatever you prefer.

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,63 @@
+brain.add_label(label_s1, borders=True)
+
+###############################################################################
+# Now we extract the pca-flipped time course from S1. Note that the most
```
So this is rather implicit, right? You could do: "...the 'pca-flipped' (see [some mne reference] for details) time course..." to brush off the obvious follow-up question: "what the heck is that"? A slightly longer explanation would have to include something about the indeterminacy of current direction in opposite banks of sulci. I do like the next sentence focusing on what HNN does, so I vote we basically leave it at this for now.

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,63 @@
+
+###############################################################################
+# Now we extract the pca-flipped time course from S1. Note that the most
+# prominent component of the median nerve response occurs in the posterior wall
+# of the central sulcus at ~0.040 sec. Since the dipolar activity here is
```
The "most prominent component" is a field pattern corresponding to an anterior-posterior dipole, right? And if it's on the posterior wall, the current direction then is deep-to-superficial, i.e., 'negative' in HNN parlance? Did I get that right?

It's quite a lot of detail to gloss over in half a sentence... If you can, feel free to use a few more words. But don't get bogged down with trying to over-simplify a complex matter.

rythorpe replied:
> The "most prominent component" is a field pattern 

I hesitate to use the term 'field pattern' because I don't want people to get EM fields mixed up with current dipoles (even though they are intimately tied), the latter of which is explicitly modeled with an inverse solution.

> corresponding to an anterior-posterior dipole, right? And if it's on the posterior wall, the current direction then is deep-to-superficial, i.e., 'negative' in HNN parlance? Did I get that right?

Kind of. You're right that this is anterior-to-posterior (i.e., backwards pointing) but this corresponds to *superficial-to-deep* current flow in the posterior wall of the CS.

> It's quite a lot of detail to gloss over in half a sentence... If you can, feel free to use a few more words. But don't get bogged down with trying to over-simplify a complex matter.

I'll see what I can do. I'm trying hard not to write a book on it here in this example but some more explanation might be good :joy:

cjayb replied:
> I hesitate to use the term 'field pattern' because I don't want people to get EM fields mixed up with current dipoles (even though they are intimately tied), the latter of which is explicitly modeled with an inverse solution.

I agree we shouldn't mention field patterns here since we don't show them! This was just to remind myself of what we're talking about at

<img width="386" alt="Screenshot 2021-01-30 at 17 04 08" src="https://user-images.githubusercontent.com/5610076/106361253-35a2df80-631d-11eb-82ea-e5eefb0c0bb7.png">


> Kind of. You're right that this is anterior-to-posterior (i.e., backwards pointing) but this corresponds to _superficial-to-deep_ current flow in the posterior wall of the CS.

Ah, yes, you're right: posterior bank, so if dipole is AP, then current direction is sup-to-deep, so negative.

cjayb commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,63 @@
 
-pick_vertex = np.argmax(np.linalg.norm(stc.data, axis=1))
+###############################################################################
+# To extract the primary response in primary somatosensory cortex (S1), we
+# create a label for the postcentral gyrus (S1) in source-space
+hemi = 'rh'
+label_tag = 'G_postcentral'
```
Hmm, ideally this would be the _sulcus_, `S_postcentral`. The visualisation is a bit harder, but it's more accurate in terms of modelling MEG signals. Is the visualisation your motivation to choose the gyrus instead?

rythorpe replied:
I tried `S_postcentral` first, however, the demarcation between 'gyrus' and 'sulcus' on the posterior wall of the CS (area 3b) is ambiguous. I found that using the 'sulcus' label did not include many of the highly responsive vertices of the inverse solution that should anatomically be considered part of area 3b.

cjayb replied:
OK, this is a detail, fine for now.

cjayb commented:
@rythorpe made some inline comments, some rather vague. Your call.

> actually looking at the CircleCI log, maybe [it's not so straightforward](https://github.com/mne-tools/mne-python/blob/main/.circleci/config.yml#L55-L66). Okay back to matplotlib. Or maybe we can skip the plot altogether and leave it as a comment like you did before. Whatever you prefer.

But it is _possible_, then? I don't think `mpl` will fly: we can't get any sort of glyph on the brain surface as it is. And even if we did, it wouldn't be possible to see deep in the folds. Plotting the inflated surface just isn't an option (we care deeply about anatomy).

If CircleCI is too cumbersome (longer spin-up times?), let's go with commenting the 3D section out, as before. WDYT? @rythorpe could you make sure that the CIs go green without it?

ntolley commented on examples/plot_simulate_somato.py:
```diff
@@ -35,18 +38,22 @@
 
+# Read and band-pass filter the raw data
 raw = mne.io.read_raw_fif(raw_fname, preload=True)
 raw.filter(1, 40)
```
```suggestion
l_freq, h_freq = 1, 40
raw.filter(l_freq, h_freq)
```

ntolley commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +129,91 @@
+# proximal drive. Note that setting ``sync_within_trial=True`` creates drives
+# with synchronous input (arriving to and transmitted by hypothetical granular
+# cells at the center of the network) to all pyramidal and basket cells that
+# receive distal drive.
```
```suggestion
# receive distal drive. Note that granule cells are not explicitly modelled within HNN. 
```

ntolley commented on examples/plot_simulate_somato.py:
```diff
@@ -94,43 +129,91 @@
+    weights_ampa=weights_ampa_d, weights_nmda=weights_nmda_d,
+    location='distal', synaptic_delays=synaptic_delays_d, seedcore=6)
+
+# Late distal input
```
```suggestion
# Late distal drive
```
Odd one out, unless you just want to keep some variety!

rythorpe commented on examples/plot_simulate_somato.py:
```diff
@@ -94,45 +136,94 @@
 # .. LINKS
 #
 # .. _MNE: https://mne.tools/
+# .. _this MNE-python example: https://mne.tools/stable/auto_examples/inverse/plot_label_source_activations.html#sphx-glr-auto-examples-inverse-plot-label-source-activations-py
```
Is there a better way to format this that works with sphinx but is PEP8 compliant?

jasmainak replied:
you can always add # noqa at the end to ask pep8 to ignore for that line (search MNE repository if you want an example)

jasmainak commented:
> But it is possible, then?

totally, some hair pulling on @rythorpe 's side potentially though ... but maybe it's not too bad with pyvista. Take a look at also the [mne-study-template CirlceCI config](https://github.com/mne-tools/mne-study-template/blob/master/.circleci/config.yml). It could be potentially simpler to follow

jasmainak commented:
@cjayb @ntolley I let one of you guys merge

rythorpe commented:
> totally, some hair pulling on @rythorpe 's side potentially though ... but maybe it's not too bad with pyvista. Take a look at also the [mne-study-template CirlceCI config](https://github.com/mne-tools/mne-study-template/blob/master/.circleci/config.yml). It could be potentially simpler to follow

I'm happy to give this a shot but it's probably not necessary for our 0.1 release. Maybe save for later?

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -94,45 +136,94 @@
+axes[1].axhline(0, c='k', ls=':', label='_nolegend_')
+axes[1].plot(1e3 * stc.times, dipole_tc, 'r--')
+average_dipoles(dpls).plot(ax=axes[1], show=False)
+axes[1].legend(['MNE vertex', 'HNN simulation'])
```
```suggestion
axes[1].legend(['MNE label average', 'HNN simulation'])
```

it's not "vertex" anymore.

jasmainak replied:
```suggestion
axes[1].legend(['MNE label average', 'HNN simulation'])
```

it's not "vertex" anymore.

rythorpe commented:
I think I address all of your comments @ntolley and @cjayb. If you're happy with the updates, feel free to merge.

rythorpe commented:
Hmm, I'm not sure what's going on with any of the CI failures. I'm able to run pytest and build the documentation on my end just fine. Any ideas @jasmainak?

jasmainak commented:
Could be some temporary issue. The MPIBackend might be fragile with respect to timings etc. I would actually stick to `JoblibBackend` for the example if we can. So we don't get stuck with MPI issues in most examples except the one meant to demo it. I restarted CircleCI. 

Did you check if you can even use multiple cores on Circle? If yes, and if MPI helps make the examples faster, we should consider having a function to set the backend globally and then set it in CircleCI so that the user does not see it in the examples and they can just run it without MPI. But anyhow, this is for 0.2

jasmainak commented:
If it fails again, try to do what I suggest above. But if you really want to stick to MPI, you can make the sphinx output verbose [here](https://github.com/jonescompneurolab/hnn-core/blob/master/doc/Makefile#L5) and then make a push so you see all the outputs not just what is coming from the error log.

cjayb reviewed (approved):
I think this looks good and reads well @rythorpe ! I have only one suggestion: consider adding a static image of the pyvista-generated plot, such as:

![3dsnapshot](https://user-images.githubusercontent.com/5610076/106361333-ac3fdd00-631d-11eb-8817-9d95db44a554.png)
 
and adding a 'caption' to it, so non-mne users get an idea of what we're doing here.

@ntolley I'm done here, please merge when @rythorpe has commented on whether he wants to add the figure (either way, great work!)

jasmainak commented:
@rythorpe if you do add a static image, please directly link to @cjayb 's image above or post a comment with the image and then link to it. That way we don't pollute the git history with images

jasmainak commented:
You can press 'i' to toggle the interface and get a cleaner image

rythorpe commented:
> @rythorpe if you do add a static image, please directly link to @cjayb 's image above or post a comment with the image and then link to it. That way we don't pollute the git history with images. It's a bit of a hack but will do the job for now

What do you think of 47e0181? I couldn't figure out how to get a formal caption to show up with an in-line figure in a sphinx-generated document, so I added more text to the paragraph prior. Also, I'm thinking that @cjayb's figure will do nicely since it looks very similar to what the user will produce should they run the code (unlike a cleaned-up version of the figure that I might create).

jasmainak commented:
I think it looks good. Do you not want to plot on the inflated surface? I can't see the label otherwise ...

rythorpe commented:
> I think it looks good. Do you not want to plot on the inflated surface? I can't see the label otherwise ...

I'm hesitant to plot this on an inflated surface because the whole point of this exercise is to prompt users to observe how real cortical structure can influence the orientation of a dipolar inverse solution.

jasmainak commented:
but you don't lose that information in inflated surface. Sulci and gyri are shown in different colors ...

rythorpe commented:
> but you don't lose that information in inflated surface. Sulci and gyri are shown in different colors ...

This might be a conversation better had over video call. In theory, you're correct. In practice, however, I'm afraid that new users won't think twice about merely plotting a distributed inverse solution on an inflated brain, selecting the most responsive vertex (or mean or principle component vertex) and orienting the time course in HNN independent of known somatotopic organization and corresponding landmarks.

jasmainak commented:
humm ... okay, let's discuss over video. I'll merge this PR when the CI comes green.

I also noticed that you used smoothing=0. I know smoothing can be confusing but it's how folks have been plotting MNE solutions for years. If you see patchy dot-like activity at a certain location, it's not because the activity there is truly 0 but rather that point may not have been sampled in the source space.

cjayb commented:
I think I'm with @rythorpe on not using the inflated surface. However, I just realised my screenshot is actually showing the anterior bank of whatever sulcus is posterior to the post-central one! Is that where the peak vertex is located? It could very well be, although physiologically the source must be on the "next sulcus over", i.e., the posterior bank of the Rolandic fissure, as we describe in the text...

This is getting complicated and potentially confusing! We shouldn't draw too much attention to this, it's supposed to be helpful.

rythorpe commented:
> I think I'm with @rythorpe on not using the inflated surface. However, I just realised my screenshot is actually showing the anterior bank of whatever sulcus is posterior to the post-central one! Is that where the peak vertex is located? It could very well be, although physiologically the source must be on the "next sulcus over", i.e., the posterior bank of the Rolandic fissure, as we describe in the text...
> 
> This is getting complicated and potentially confusing! We shouldn't draw too much attention to this, it's supposed to be helpful.

@cjayb since our method extracts the primary principle component time course from the post-central gyrus label, it's hard to tell exactly where the source we model is located. From the literature, you're right that it should be somewhere in area 3b/1 of the hand representation (i.e., the omega-shaped 'ripple' in the CS). I know it's late for you, but I invited you to @jasmainak and my quick zoom call later today if you wish to join.

jasmainak commented:
I'm actually free now if you guys want to talk earlier :)

On Mon, Feb 1, 2021 at 3:49 PM Ryan Thorpe <notifications@github.com> wrote:

> I think I'm with @rythorpe <https://github.com/rythorpe> on not using the
> inflated surface. However, I just realised my screenshot is actually
> showing the anterior bank of whatever sulcus is posterior to the
> post-central one! Is that where the peak vertex is located? It could very
> well be, although physiologically the source must be on the "next sulcus
> over", i.e., the posterior bank of the Rolandic fissure, as we describe in
> the text...
>
> This is getting complicated and potentially confusing! We shouldn't draw
> too much attention to this, it's supposed to be helpful.
>
> @cjayb <https://github.com/cjayb> since our method extracts the primary
> principle component time course from the post-central gyrus label, it's
> hard to tell exactly where the source we model is located. From the
> literature, you're right that it should be somewhere in area 3b/1 of the
> hand representation (i.e., the omega-shaped 'ripple' in the CS). I know
> it's late for you, but I invited you to @jasmainak
> <https://github.com/jasmainak> and my quick zoom call later today if you
> wish to join.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jonescompneurolab/hnn-core/pull/253#issuecomment-771148147>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADY6FISTW4JGLP3BZ2UAVWLS44HWJANCNFSM4WN44E4A>
> .
>

rythorpe commented:
How about these two images showing the label of interest and MNE activation, respectively.

![fig_1](https://user-images.githubusercontent.com/20212206/106524603-cfe75c80-64b0-11eb-9607-3415195c3e7a.png)

![fig_2](https://user-images.githubusercontent.com/20212206/106524542-b514e800-64b0-11eb-835e-497454e75eb9.png)

jasmainak commented:
I like it!

Mainak

On Mon, Feb 1, 2021 at 5:14 PM Ryan Thorpe <notifications@github.com> wrote:

> How about these two images showing the label of interest and MNE
> activation, respectively.
>
> [image: fig_1]
> <https://user-images.githubusercontent.com/20212206/106524603-cfe75c80-64b0-11eb-9607-3415195c3e7a.png>
>
> [image: fig_2]
> <https://user-images.githubusercontent.com/20212206/106524542-b514e800-64b0-11eb-835e-497454e75eb9.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jonescompneurolab/hnn-core/pull/253#issuecomment-771196160>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADY6FIQAMUCNT44WCRXBD33S44RTRANCNFSM4WN44E4A>
> .
>

rythorpe commented:
I'm learning that formatting with reStructured text can be frustrating. I've updated the code to produce the figures shown in my [previous comment](https://github.com/jonescompneurolab/hnn-core/pull/253#issuecomment-771196160) and updated the corresponding links, though I'm open to suggestions for improving the formatting.

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -94,45 +143,99 @@
+# Now we run the simulation over 2 trials so that we can plot the average
+# aggregate dipole. As in :ref:`the MPIBackend example
+# <sphx_glr_auto_examples_plot_simulate_mpi_backend.py>`, we can use
+# ``MPIBackend`` to reduce the simulation time by parallizing across cells in
```
I was going to merge but then noticed there are still references to `MPIBackend` :)

rythorpe replied:
Good catch. Fixed in https://github.com/jonescompneurolab/hnn-core/pull/253/commits/1d6e6391894eb181947afc9c6c1b6b5a09bd915d.

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,77 @@
+# ``pyvista`` python library). You should get 2 plots, the first showing the
+# post-central gyrus label from which the dipole time course was extracted and
+# the second showing MNE activation at 0.040 sec that resemble the following
+# images:
```
add a blank line after this so the image goes into another paragraph

jasmainak commented on examples/plot_simulate_somato.py:
```diff
@@ -57,35 +64,77 @@
+# post-central gyrus label from which the dipole time course was extracted and
+# the second showing MNE activation at 0.040 sec that resemble the following
+# images:
+# |mne_label_fig|
+# |mne_activity_fig|
```
actually if you put this after the commented lines, it would look most natural because the plot should appear after the code. But I won't insist if you disagree ...

jasmainak commented:
Thank you @rythorpe ! It took longer than expected but final product looks great!

cjayb commented:
Looks amazing, really great touch with the figures!!

agramfort commented on examples/plot_simulate_somato.py:
```diff
@@ -94,45 +146,96 @@
+# .. _this MNE-python example: https://mne.tools/stable/auto_examples/inverse/plot_label_source_activations.html#sphx-glr-auto-examples-inverse-plot-label-source-activations-py # noqa
+# .. |mne_label_fig| image:: https://user-images.githubusercontent.com/20212206/106524603-cfe75c80-64b0-11eb-9607-3415195c3e7a.png # noqa
+#   :width: 400
+# .. |mne_activity_fig| image:: https://user-images.githubusercontent.com/20212206/106524542-b514e800-64b0-11eb-835e-497454e75eb9.png # noqa
```
you do this as you don't want to have to setup pyvista + vtk on your CI?

jasmainak replied:
yes it's not really worth the effort to set up headless 3D rendering at this stage of HNN development cycle. We might do it after the 0.1 release though (which will happen soon)

rythorpe replied:
Out of curiosity, what is the primary difficulty with setting up the 3D rendering packages on CI?

cjayb replied:
OpenGL requires hardware (GPU) which the headless servers don't have/expose. A software emulation layer is needed, which I'm guessing for something as complex as VTK can be a bit of a hassle. Could also be a performance-issue. It's certainly doable (look at mne!), but I'm glad we decided to side-step it for now: there's plenty to work on as it is :)
